lib/SILOptimizer/Utils/CanonicalizeBorrowScope.cpp

```diff
@@ -224,21 +224,34 @@ bool CanonicalizeBorrowScope::visitBorrowScopeUses(SILValue innerValue,
     // Gather the uses before updating any of them.
     // 'value' may be deleted in this loop after rewriting its last use.
     // 'use' may become invalid after processing its user.
+
     SmallVector<Operand *, 4> uses(value->getUses());
-    for (Operand *use : uses) {
+    unsigned index = 0;
+    while (index < uses.size()) {
+      SWIFT_DEFER { ++index; };
+      auto *use = uses[index];
       auto *user = use->getUser();
       // Incidental uses, such as debug_value may be deleted before they can be
       // processed. Their user will now be nullptr. This means that value
       // is dead, so just bail.
       if (!user)
         break;
 
-      // Recurse through copies.
+      // Recurse through copies and moves.
       if (auto *copy = dyn_cast<CopyValueInst>(user)) {
         if (!isPersistentCopy(copy)) {
           defUseWorklist.insert(copy);
           continue;
         }
+      } else if (auto *move = dyn_cast<MoveValueInst>(user)) {
+        if (!move->isLexical() || innerValue->isLexical()) {
+          for (auto *moveUse : move->getUses()) {
+            uses.push_back(moveUse);
+          }
+          move->replaceAllUsesWith(move->getOperand());
+          deleter.forceDelete(move);
```

Comment:
It's generally bad form to mutate SIL in the middle of an analysis--confusing/bug-prone from the client's perspective. Can't the analysis just scope the moves and let the rewriter visitors actually delete them?

Reply:
Fixed to just add the move to the worklist.